priv/repo/migrations/20200316103722_create_user_auth_tables.exs

```diff
@@ -0,0 +1,26 @@
+defmodule Demo.Repo.Migrations.CreateUserAuthTables do
+  use Ecto.Migration
+
+  def change do
+    execute "CREATE EXTENSION IF NOT EXISTS citext", ""
+
+    create table(:users) do
+      add :email, :citext, null: false
+      add :hashed_password, :string, null: false
+      add :confirmed_at, :naive_datetime
+      timestamps()
+    end
+
+    create unique_index(:users, [:email])
+
+    create table(:user_tokens) do
+      add :user_id, references(:users, on_delete: :delete_all), null: false
+      add :token, :binary, null: false
+      add :context, :string, null: false
+      add :sent_to, :string
+      add :inserted_at, :naive_datetime
+    end
+
+    create unique_index(:user_tokens, [:context, :token])
```

Comment:
@josevalim When running this migration against mysql, I'm getting the following error when creating the `user_tokens_context_token_index`:

```
** (MyXQL.Error) (1170) (ER_BLOB_KEY_WITHOUT_LENGTH) BLOB/TEXT column 'token' used in key specification without a key length
    (ecto_sql 3.4.1) lib/ecto/adapters/sql.ex:612: Ecto.Adapters.SQL.raise_sql_call_error/1
    (elixir 1.10.2) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (ecto_sql 3.4.1) lib/ecto/adapters/sql.ex:705: Ecto.Adapters.SQL.execute_ddl/4
    (ecto_sql 3.4.1) lib/ecto/migration/runner.ex:343: Ecto.Migration.Runner.log_and_execute_ddl/3
    (ecto_sql 3.4.1) lib/ecto/migration/runner.ex:117: anonymous fn/6 in Ecto.Migration.Runner.flush/0
    (elixir 1.10.2) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto_sql 3.4.1) lib/ecto/migration/runner.ex:116: Ecto.Migration.Runner.flush/0
    (stdlib 3.10) timer.erl:166: :timer.tc/1
```

I had a similar error on `:email`, but I switched the type from `:text` to `:string` and was able to get the users index working. Do you happen to know what we need to do here to keep mysql happy?

Reply:
We need to add a size to both columns. The email one can be the same size as we validate. For token, the size is 32! Can you please try that?

Reply:
Perfect, that worked! Here's my current mysql version of the migration:

```elixir
defmodule DemoMysql.Repo.Migrations.CreateUserAuthTables do
  use Ecto.Migration

  def change do
    create table(:users) do
      add :email, :"varchar(80)", null: false
      add :hashed_password, :string, null: false
      add :confirmed_at, :naive_datetime
      timestamps()
    end

    create unique_index(:users, [:email])

    create table(:user_tokens) do
      add :user_id, references(:users, on_delete: :delete_all), null: false
      add :token, :"binary(32)", null: false
      add :context, :string, null: false
      add :sent_to, :string
      add :inserted_at, :naive_datetime
    end

    create unique_index(:user_tokens, [:context, :token])
  end
end
```

Reply:
You can do size: 32 and size: 80 instead and it should be compatible with all DBs. :)

Reply:
Oh, it should be size: 160 for email. :)

Reply:
Thanks for the tip on the `:size` attribute! :heart: It works well for the email field, but unfortunately when I use it with `:binary`, it generates a `tinyblob` field instead of `binary(32)`.

Reply:
@aaronrenner tinyblob should be fine. Ecto doesn't emit it, so I am thinking MySQL treats them as equivalent. Or are you running into errors?

Reply:
maybe we should just go with the :string type (which still can have a size) then

Reply:
@wojtekmach I just tried the following
```elixir
add :token, :string, null: false, size: 32
```

And got the following error when running the test suite:
```

       1) test deliver_update_email_instructions/3 sends token through notification (DemoMysql.AccountsTest)
          test/demo_mysql/accounts_test.exs:164
          ** (MyXQL.Error) (1366) (ER_TRUNCATED_WRONG_VALUE_FOR_FIELD) Incorrect string value: '\xCE\xDC\xF9\x1D\x98\x05...' for column 'token' at row 1
```

@josevalim when I define it as a `:binary`
```elixir
add :token, :binary, null: false, size: 32
```
I get the following error when running the `create_unique_index` piece of the migration
```
** (MyXQL.Error) (1170) (ER_BLOB_KEY_WITHOUT_LENGTH) BLOB/TEXT column 'token' used in key specification without a key length
    (ecto_sql 3.4.1) lib/ecto/adapters/sql.ex:612: Ecto.Adapters.SQL.raise_sql_call_error/1
    (elixir 1.10.2) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (ecto_sql 3.4.1) lib/ecto/adapters/sql.ex:705: Ecto.Adapters.SQL.execute_ddl/4
    (ecto_sql 3.4.1) lib/ecto/migration/runner.ex:343: Ecto.Migration.Runner.log_and_execute_ddl/3
    (ecto_sql 3.4.1) lib/ecto/migration/runner.ex:117: anonymous fn/6 in Ecto.Migration.Runner.flush/0
    (elixir 1.10.2) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto_sql 3.4.1) lib/ecto/migration/runner.ex:116: Ecto.Migration.Runner.flush/0
    (stdlib 3.10) timer.erl:166: :timer.tc/1
```
When I check the database, I saw `token` is defined as a `TINYBLOB`.

Finally, when I define it as `binary(32)` everything works and the tests pass.
```elixir
add :token, :"binary(32)", null: false
```
However, if I do it may not work with other databases like SQL server. What you you think?

Reply:
It is an Ecto bug. If binary with a size is given, we should not emit blob, but rather varbinary. I will fix it tomorrow.

Reply:
EctoSQL v3.4.2 is out. It should work now!

Reply:
Thanks, so much! That works!

I did notice I can't put `size: 32` on the postgres version of the migration, 
```elixir
add :token, :binary, null: false, size: 32
```
because I'll get the following error:
```
** (Postgrex.Error) ERROR 42601 (syntax_error) type modifier is not allowed for type "bytea"
    (ecto_sql 3.4.2) lib/ecto/adapters/sql.ex:612: Ecto.Adapters.SQL.raise_sql_call_error/1
    (elixir 1.10.2) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (ecto_sql 3.4.2) lib/ecto/adapters/sql.ex:705: Ecto.Adapters.SQL.execute_ddl/4
    (ecto_sql 3.4.2) lib/ecto/migration/runner.ex:343: Ecto.Migration.Runner.log_and_execute_ddl/3
    (ecto_sql 3.4.2) lib/ecto/migration/runner.ex:117: anonymous fn/6 in Ecto.Migration.Runner.flush/0
    (elixir 1.10.2) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto_sql 3.4.2) lib/ecto/migration/runner.ex:116: Ecto.Migration.Runner.flush/0
    (stdlib 3.10) timer.erl:166: :timer.tc/1
```
Is this expected and ok that I don't include `:size` on the postgres version?

Reply:
Soundsd good. No need to include it for PG.